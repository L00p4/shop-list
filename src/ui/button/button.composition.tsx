import Button, { ButtonSkeleton } from '.'

export const ButtonComprar = () => <Button variant="primary">Comprar</Button>

export const Skeleton = () => <ButtonSkeleton />
