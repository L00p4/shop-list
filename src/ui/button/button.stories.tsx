import { Meta, StoryObj } from '@storybook/react'

import Button, { ButtonSkeleton } from '.'

const meta: Meta<typeof Button> = {
  title: 'UI/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    variant: {
      control: { type: 'select' },
      options: ['primary', 'secondary', 'danger']
    },
    size: {
      control: { type: 'select' },
      options: ['small', 'medium', 'large']
    },
    disabled: {
      control: { type: 'boolean' }
    }
  },
  args: {
    children: 'Button',
    variant: 'primary',
    size: 'medium',
    disabled: false
  }
}

type Story = StoryObj<typeof Button>

// Primary variant stories
export const Primary: Story = {
  args: {
    children: '🛒 Comprar',
    variant: 'primary'
  }
}

export const Secondary: Story = {
  args: {
    children: '📋 Editar',
    variant: 'secondary'
  }
}

export const Danger: Story = {
  args: {
    children: '🗑️ Remover',
    variant: 'danger'
  }
}

// Size variations
export const Compact: Story = {
  args: {
    children: 'Compact',
    size: 'compact'
  }
}

export const Small: Story = {
  args: {
    children: '+ Add',
    size: 'small'
  }
}

export const Medium: Story = {
  args: {
    children: 'Adicionar ao Carrinho',
    size: 'medium'
  }
}

export const Large: Story = {
  args: {
    children: '✅ Finalizar Compra',
    size: 'large'
  }
}

// State variations
export const Disabled: Story = {
  args: {
    children: '🔒 Finalizar (itens pendentes)',
    disabled: true
  }
}

// Real usage examples from our wireframes
export const AddToCartButton: Story = {
  args: {
    children: '+',
    variant: 'primary',
    size: 'small'
  },
  parameters: {
    docs: {
      description: {
        story: 'Botão para adicionar item ao carrinho na lista de compras'
      }
    }
  }
}

export const EditButton: Story = {
  args: {
    children: '✏️',
    variant: 'secondary',
    size: 'small'
  },
  parameters: {
    docs: {
      description: {
        story: 'Botão para editar item na lista'
      }
    }
  }
}

export const NewListButton: Story = {
  args: {
    children: '+ Nova',
    variant: 'primary',
    size: 'small'
  },
  parameters: {
    docs: {
      description: {
        story: 'Botão para criar nova lista na tela inicial'
      }
    }
  }
}

export const ConfirmPurchase: Story = {
  args: {
    children: '✅ Confirmar Compra',
    variant: 'primary',
    size: 'large'
  },
  parameters: {
    docs: {
      description: {
        story: 'Botão principal para finalizar compra no modal de conferência'
      }
    }
  }
}

// All variants showcase
export const AllVariants: Story = {
  render: () => (
    <div style={{ display: 'flex', gap: 'var(--space-4)', flexWrap: 'wrap' }}>
      <Button variant="primary">Primary</Button>
      <Button variant="secondary">Secondary</Button>
      <Button variant="danger">Danger</Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Todas as variantes do botão lado a lado'
      }
    }
  }
}

// All sizes showcase
export const AllSizes: Story = {
  render: () => (
    <div
      style={{
        display: 'flex',
        gap: 'var(--space-4)',
        alignItems: 'center',
        flexWrap: 'wrap'
      }}
    >
      <Button size="compact">Compact</Button>
      <Button size="small">Small</Button>
      <Button size="medium">Medium</Button>
      <Button size="large">Large</Button>
    </div>
  ),
  parameters: {
    docs: {
      description: {
        story: 'Todos os tamanhos do botão lado a lado'
      }
    }
  }
}

export const Skeleton: Story = {
  render: () => <ButtonSkeleton />,
  args: {}
}

export default meta
