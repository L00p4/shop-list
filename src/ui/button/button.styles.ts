// Button/button.styles.ts
import styled, { css } from 'styled-components'
import { ButtonProps } from '.'

type WrapperButtonProps = {
  variant: ButtonProps['variant']
  size: ButtonProps['size']
}

const buttonVariants = {
  primary: css`
    background: var(--purple-500);
    color: var(--text-primary);
    border: 1px solid var(--purple-500);

    &:hover:not(:disabled) {
      background: var(--purple-700);
      border-color: var(--purple-700);
    }

    &:active:not(:disabled) {
      background: var(--purple-700);
      transform: translateY(1px);
    }
  `,
  secondary: css`
    background: var(--border-default);
    color: var(--text-secondary);
    border: 1px solid var(--border-default);

    &:hover:not(:disabled) {
      background: var(--border-default);
      border-color: var(--text-tertiary);
    }

    &:active:not(:disabled) {
      background: var(--border-default);
      transform: translateY(1px);
    }
  `,
  danger: css`
    background: var(--error);
    color: var(--text-primary);
    border: 1px solid var(--error);

    &:hover:not(:disabled) {
      background: var(--error-light);
      border-color: var(--error-light);
    }

    &:active:not(:disabled) {
      background: var(--error-light);
      transform: translateY(1px);
    }
  `,
  warning: css`
    background: var(--warning);
    color: var(--text-primary);
    border: 1px solid var(--warning);

    &:hover:not(:disabled) {
      background: var(--warning-light);
      border-color: var(--warning-light);
    }

    &:active:not(:disabled) {
      background: var(--warning-light);
      transform: translateY(1px);
    }
  `
}

const buttonSizes = {
  compact: css`
    padding: var(--space-1);
    font-size: var(--font-size-xs);
    border-radius: var(--radius-sm);
  `,
  small: css`
    padding: var(--space-1) var(--space-3);
    font-size: var(--font-size-xs);
    border-radius: var(--radius-md);
  `,
  medium: css`
    padding: var(--space-2) var(--space-4);
    font-size: var(--font-size-sm);
    border-radius: var(--radius-md);
  `,
  large: css`
    padding: var(--space-3) var(--space-6);
    font-size: var(--font-size-base);
    border-radius: var(--radius-lg);
  `
}

export const WrapperButton = styled.button<WrapperButtonProps>`
  /* Reset and base styles */
  border: none;
  outline: none;
  cursor: pointer;
  font-family: inherit;
  font-weight: var(--font-medium);

  /* Layout */
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: var(--space-2);

  /* Transitions */
  transition: all var(--transition-normal);

  /* Apply variant styles */
  ${({ variant = 'primary' }) => buttonVariants[variant]}

  /* Apply size styles */
  ${({ size = 'medium' }) => buttonSizes[size]}

  /* Disabled state */
  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
    transform: none;
  }

  /* Focus styles for accessibility */
  &:focus-visible {
    outline: 2px solid var(--purple-500);
    outline-offset: 2px;
  }

  /* Prevent text selection */
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;

  /* Touch improvements for mobile */
  touch-action: manipulation;
  -webkit-tap-highlight-color: transparent;
`
