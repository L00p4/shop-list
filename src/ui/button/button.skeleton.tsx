import { WrapperButton } from './button.styles'

const ButtonSkeleton = () => (
  <WrapperButton size="medium" variant="primary" disabled>
    <h1 className="skeleton">Button</h1>
  </WrapperButton>
)

export default ButtonSkeleton
