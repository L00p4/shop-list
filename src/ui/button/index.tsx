import React from 'react'

import { WrapperButton } from './button.styles'

export { default as ButtonSkeleton } from './button.skeleton'

export type ButtonProps = {
  children: React.ReactNode
  variant?: 'primary' | 'secondary' | 'danger' | 'warning'
  size?: 'compact' | 'small' | 'medium' | 'large'
  disabled?: boolean
  onClick?: () => void
  type?: 'button' | 'submit' | 'reset'
  className?: string
}

const Button: React.FC<ButtonProps> = ({
  children,
  variant = 'primary',
  size = 'medium',
  disabled = false,
  onClick,
  type = 'button',
  className,
  ...rest
}) => (
  <WrapperButton
    variant={variant}
    size={size}
    disabled={disabled}
    onClick={onClick}
    type={type}
    className={className}
    {...rest}
  >
    {children}
  </WrapperButton>
)

export default Button

// Button/index.tsx
// import React from 'react'
// import { WrapperButton } from './button.styles'

// export { default as ButtonSkeleton } from './button.skeleton'

// export interface ButtonProps
//   extends React.ButtonHTMLAttributes<HTMLButtonElement> {
//   children: React.ReactNode
//   variant?: 'primary' | 'secondary' | 'danger'
//   size?: 'small' | 'medium' | 'large'
//   isLoading?: boolean
//   leftIcon?: React.ReactNode
//   rightIcon?: React.ReactNode
//   fullWidth?: boolean
// }

// const Button: React.FC<ButtonProps> = ({
//   children,
//   variant = 'primary',
//   size = 'medium',
//   disabled = false,
//   isLoading = false,
//   leftIcon,
//   rightIcon,
//   fullWidth = false,
//   className,
//   ...rest
// }) => {
//   const isDisabled = disabled || isLoading

//   return (
//     <WrapperButton
//       variant={variant}
//       size={size}
//       disabled={isDisabled}
//       fullWidth={fullWidth}
//       className={className}
//       {...rest}
//     >
//       {isLoading ? (
//         <>
//           <span className="animate-pulse">‚è≥</span>
//           {children}
//         </>
//       ) : (
//         <>
//           {leftIcon && <span data-testid="left-icon">{leftIcon}</span>}
//           {children}
//           {rightIcon && <span data-testid="right-icon">{rightIcon}</span>}
//         </>
//       )}
//     </WrapperButton>
//   )
// }

// export default Button
