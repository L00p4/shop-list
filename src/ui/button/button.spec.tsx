import '@testing-library/jest-dom'
import 'jest-environment-jsdom'
import 'jest-styled-components'

import { render } from '@testing-library/react'

import { ButtonComprar, Skeleton } from './button.composition'

describe('<Button />', () => {
  it('should render the heading', () => {
    render(<ButtonComprar />)
  })

  it('should render Primary', () => {
    expect(render(<ButtonComprar />).container).toMatchSnapshot()
  })

  it('should render Skeleton', () => {
    expect(render(<Skeleton />).container).toMatchSnapshot()
  })
})
