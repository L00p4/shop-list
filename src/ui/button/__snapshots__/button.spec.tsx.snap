// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Button /> should render Primary 1`] = `
.c0 {
  border: none;
  outline: none;
  cursor: pointer;
  font-family: inherit;
  font-weight: var(--font-medium);
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: var(--space-2);
  transition: all var(--transition-normal);
  background: var(--purple-500);
  color: var(--text-primary);
  border: 1px solid var(--purple-500);
  padding: var(--space-2) var(--space-4);
  font-size: var(--font-size-sm);
  border-radius: var(--radius-md);
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  touch-action: manipulation;
  -webkit-tap-highlight-color: transparent;
}

.c0:hover:not(:disabled) {
  background: var(--purple-700);
  border-color: var(--purple-700);
}

.c0:active:not(:disabled) {
  background: var(--purple-700);
  transform: translateY(1px);
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
  transform: none;
}

.c0:focus-visible {
  outline: 2px solid var(--purple-500);
  outline-offset: 2px;
}

<div>
  <button
    class="c0"
    type="button"
    variant="primary"
  >
    Comprar
  </button>
</div>
`;

exports[`<Button /> should render Skeleton 1`] = `
.c0 {
  border: none;
  outline: none;
  cursor: pointer;
  font-family: inherit;
  font-weight: var(--font-medium);
  display: inline-flex;
  align-items: center;
  justify-content: center;
  gap: var(--space-2);
  transition: all var(--transition-normal);
  background: var(--purple-500);
  color: var(--text-primary);
  border: 1px solid var(--purple-500);
  padding: var(--space-2) var(--space-4);
  font-size: var(--font-size-sm);
  border-radius: var(--radius-md);
  user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  touch-action: manipulation;
  -webkit-tap-highlight-color: transparent;
}

.c0:hover:not(:disabled) {
  background: var(--purple-700);
  border-color: var(--purple-700);
}

.c0:active:not(:disabled) {
  background: var(--purple-700);
  transform: translateY(1px);
}

.c0:disabled {
  opacity: 0.6;
  cursor: not-allowed;
  transform: none;
}

.c0:focus-visible {
  outline: 2px solid var(--purple-500);
  outline-offset: 2px;
}

<div>
  <button
    class="c0"
    disabled=""
    variant="primary"
  >
    <h1
      class="skeleton"
    >
      Button
    </h1>
  </button>
</div>
`;
